### DISCUSSION TITLE
What is "SRG_PerPass slots array or structure"?

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#14114] [Help Wanted]
https://development--o3deorg.netlify.app/docs/atom-guide/dev-guide/passes/pass-template-file-spec/#slots

Thanks.

**Comment by VickyAtAZ** [comment]
Each pass may have a pass SRG (Shader Resource Group). A pass SRG contains data can be used by any draw items which are submitted in this pass. 

The pass SRG is specified via pass's PassData structure. 
For example, the Forward pass has a pass SRG which is specified in `o3de\Gems\Atom\Feature\Common\Assets\Passes\OpaqueParent.pass` file where the PassRequest creates Forward pass from ForwardPassTemplate
 ```
   "PassData": {
                        "$type": "RasterPassData",
                        "DrawListTag": "forwardWithSubsurfaceOutput",
                        "PipelineViewTag": "MainCamera",
                        "PassSrgShaderAsset": {
                            "FilePath": "Shaders/ForwardPassSrg.shader"
                        }
```
The PassSrgShaderAsset is "Shaders/ForwardPassSrg.shader" is using an azsl which contains the definition of the pass srg in 
in `\o3de\Gems\Atom\Feature\Common\Assets\ShaderLib\Atom\Features\Pipeline\Forward\ForwardPassSrg.azsli` file. 

And in `\o3de\Gems\Atom\Feature\Common\Assets\Passes\Forward.pass` file, each slot is mapping to a parameter defined in `ForwardPassSrg.azsli`
For example, this slot
```
{
                    "Name": "FullscreenShadow",
                    "ShaderInputName": "m_fullscreenShadow",
                    "SlotType": "Input",
                    "ScopeAttachmentUsage": "Shader"
}
```
is mapping to `Texture2D<float> m_fullscreenShadow;` defined in 'ShaderResourceGroup PassSrg : SRG_PerPass'

It's possible a pass SRG may contain a fixed array too. And a pass slot can map to one element of the array as well. 
You can find an example in `\o3de\Gems\Atom\Feature\Common\Assets\Shaders\Checkerboard\CheckerboardColorResolveCS.azsl` which `Texture2DMS<float4> m_color[MaxNumberResolveTexturesDouble];` is a texture array. And in `\o3de\Gems\Atom\Feature\Common\Assets\Passes\CheckerboardResolveColor.pass` file, this slot is mapping to the 5th item in the array

```
                {
                    "Name": "InputColor2_curr",
                    "SlotType": "Input",
                    "ShaderInputName": "m_color",
                    "ShaderInputArrayIndex": "4",
                    "ScopeAttachmentUsage": "Shader",
                    "LoadStoreAction": {
                        "LoadAction": "Load",
                        "StoreAction": "DontCare",
                        "LoadActionStencil": "DontCare",
                        "StoreActionStencil": "DontCare"
                    }
                },
```
